### PR TITLE
[Feature 6] Matching Algorithm

### LINKED ISSUE #2: [Feature 6] Matching Algorithm
A system that will update the user's compatibility with the groups in the repository.

### PR DESCRIPTION
**Pull request by helenaglow**
**UPDATE: As per Hannah's advice, I have made changes!

The changes can be seen under the "Updated matching algorithm entities commit"! I am still planning on adding UserMatchesFactory interface and more in regard to the Factory design.

Rather than having the sorter, I am using comparable for the new class GroupScore. More information is added on the code.

Like before, there are still errors as I am under the assumption that getPublicProfile (for User) and getPreferences() for both User and Group will be added.

### CODE REVIEW

File: src/main/java/Entities/UserMatches/SelectionSortMatchGroups.java
Comment (hannahpan1) [inline src/main/java/Entities/UserMatches/SelectionSortMatchGroups.java]: For this I think we can use the comparable interface (since we are authoring the class we can use this instead of the Comparator interface) and the CompareTo instead of making our own sort interface. This will allow you to use the collections sort method as well as all other methods that can work with the Comparable interface. In comparison if we make our sort interface we can not access all the methods that know how to work with Comparable items. This would work if we have like the the match class implement comparable.

File: src/main/java/Entities/UserMatches/SelectionSortMatchGroups.java
Comment (hannahpan1) [inline src/main/java/Entities/UserMatches/SelectionSortMatchGroups.java]: If we use the comparable interface then we won't have to specify the type of the list like List\<Object>, which seems to be an issue we are running into when you call sort on a list of groups instead of a list of objects (on line 34 of UserMatches.java).

File: src/main/java/Entities/UserMatches/UserMatches.java
Comment (hannahpan1) [inline src/main/java/Entities/UserMatches/UserMatches.java]: Instead of having the match method put the similarity score and and group into a hash map, as mentioned before we could have a list of matches, where match is it's own class that has a similarity score instance attribute (this can be defined in the constructor the same way that calculateSimilarity() method is defining the similarity score). This match could implement comparable and then we can just make a list of match instances for each group in this method. Then we can just sort the list of matches using Collections' sort method. I think this will follow the clean architecture principle of single responsibility better and also get rid of some methods.

**Review by hannahpan1 — COMMENTED** [review]
Great first implementation of your code! Wrote some comments about the implementation that I think might help simplify the code a bit.

**Review by ileuxali — COMMENTED** [review]
I'll help make changes in user entity :^)

**Review by ileuxali — DISMISSED** [review]

**Review by aaryavatsa — DISMISSED** [review]
This looks good! Thank you for the nicely formatted documentation, it was very easy to follow.

**Comment by helenaglow** [comment]
I have uploaded the Matching Algorithm Use Case (without screens and tests) and updated Matching Algorithm Entities.

There some changes I will likely have to make depending on if we do serialization or csv.

Furthermore, at the moment there are two files I do not think I will need: MatchingAlgorithmInputBoundary and MatchingAlgorithmRequestModel. This is because the use case does not really have input. Currently I have booleans for whether the "Refresh" or "Find Group" buttons are clicked. However, I do not think this is necessary. Instead, in the screens, if those buttons are clicked the controller will be called. So, I do not think these files are necessary, but I will confirm with TA. 

Do you guys think I should make a UserMatches and GroupScore factory? I am not sure they are necessary, but I am making multiple instances of GroupScore. So I am not 100% sure.

**_I had another note but I cannot think of it at the moment. I will edit the message when I remember it :(_

**Review by ileuxali — DISMISSED** [review]

File: src/main/java/Entities/UserMatches/GroupScore.java
Comment (soheegoo) [inline src/main/java/Entities/UserMatches/GroupScore.java]: Nice description! This makes it really easy to understand and follow your code. Thanks for making the such clear comments :)

File: src/main/java/Entities/UserMatches/GroupScore.java
Comment (soheegoo) [inline src/main/java/Entities/UserMatches/GroupScore.java]: Really smart idea to implement the Comparable interface!

File: src/main/java/Entities/UserMatches/GroupScore.java
Comment (soheegoo) [inline src/main/java/Entities/UserMatches/GroupScore.java]: Maybe add some javadoc comments on how this function works to make it clear for the user

File: src/main/java/Entities/UserMatches/MatchingAlgorithm.java
Comment (soheegoo) [inline src/main/java/Entities/UserMatches/MatchingAlgorithm.java]: Good use of clean architecture incase we want to introduce different matches in the future!

File: src/main/java/matching_algorithm_use_case/MatchingAlgorithmResponseModel.java
Comment (soheegoo) [inline src/main/java/matching_algorithm_use_case/MatchingAlgorithmResponseModel.java]: Just a general note, I think you should write your input variables as like updateMatches to follow convention and make it easier for whoever is reading.

**Review by soheegoo — DISMISSED** [review]

File: src/main/java/Entities/UserMatches/GroupScore.java
Comment (aaryavatsa) [reply src/main/java/Entities/UserMatches/GroupScore.java]: Agree with Sohee ^ setting the bar for doc strings 🥇  woooo

**Review by aaryavatsa — COMMENTED** [review]
Everything looks good and I think you should be ready to merge -- but why is the autograder failing for this PR? It says there is a compliation error. Just glancing at it, I think it may be a problem with some of your method calls for the group leader. I think you should take a look at the autograder and the compilation error before we merge this.

**Comment by helenaglow** [comment]
> Everything looks good and I think you should be ready to merge -- but why is the autograder failing for this PR? It says there is a compliation error. Just glancing at it, I think it may be a problem with some of your method calls for the group leader. I think you should take a look at the autograder and the compilation error before we merge this.

@aaryavatsa good catch it seems that there is an error with the group entity itself, but I didn't commit the group entity? (I don't use group leader anywhere in my code). 

 Also, there are is an error in group score as it is dependent on some methods and parts of Group and Group Public Profile that have not been added yet... not sure what I should do in this situation.

**Comment by helenaglow** [comment]
I accidently kept committing without pushing. My bad.

Updates:
- I have added Java docs
- clarified the Request Model, will now be requesting User's username
- Changed UserMatches entites so that groups that the user is in or are not for the preferred course code are cut from the matches
- improved variable naming as Sohee suggested
- Presenter is now OutputBoundary

My code is still dependent on @soheegoo's .getGroupMembersUsernames() & .getGroupProfile(), and @hannahpan1's and @julia80887's .getPreferences() & .getCourseCode(). We already discussed and should be implemented with their coming pull requests

Furthermore, some more updates from my last pull request. I discussed with the TA at the lab about the implementation of factories. They said it was helpful when entities are rather complex, and it is up to me to decide what I deem to be necessary. I do not think a factory would be necessary for some reasons:

- UserMatches takes in an instance of User: this means that I would not have to make different cases of UserMatches based on the type of user. All match requirements are part of the User interface.
- I only make one UserMatches in the interactor
- It is not super complex


Please feel free to leave any comments or suggestions :)

File: src/main/java/Entities/UserMatches/MatchingAlgorithm.java
Comment (hannahpan1) [inline src/main/java/Entities/UserMatches/MatchingAlgorithm.java]: This is a great idea! We can also use this to add different ways to match items in the future.

File: src/main/java/Entities/UserMatches/UserMatches.java
Comment (hannahpan1) [inline src/main/java/Entities/UserMatches/UserMatches.java]: Looks good! One thing to note is that in my most recent PR I have made course code an instance variable as well for the user public profile.  This means for cutByCourseCodeAndMemberList() method, instead of using the get method for preferences you can just look up course code.

Comment (hannahpan1) [reply src/main/java/Entities/UserMatches/UserMatches.java]: Also why do we get rid of group score in getMatches() when to turn them into a list of strings for the interactor we use getGroup anyways?

File: src/main/java/matching_algorithm_use_case/MatchingAlgorithmOutputBoundary.java
Comment (hannahpan1) [inline src/main/java/matching_algorithm_use_case/MatchingAlgorithmOutputBoundary.java:10]: Does this run without error? The prepareFailView method is returning a MatchingAlgorithmResponseModel, which is not a string to my knowledge.

**Review by hannahpan1 — CHANGES_REQUESTED** [review]
Hi! Everything looks good. I have a couple of questions I left in the review as well as one small change that I think needs to be implemented for our use cases to link together.

**Comment by helenaglow** [comment]
dsfdfsdfsf

**Comment by helenaglow** [comment]
I have committed changes to the my entities based on Hannah's request and also updated my Interactor (which I found a small error in and a code smell 🤢). Also aligned my DsGateWay with Leo,s pull request

**Review by ipekakyol — DISMISSED** [review]
The javadocs make it really easy to follow the code. I do not see any clean architecture violations and everything looks good.

File: src/main/java/entities/user_matches/GroupScore.java
Comment (hannahpan1) [inline src/main/java/entities/user_matches/GroupScore.java]: Does this import statement work? To my knowledge our entities package is capitalized. This seems also to be the problem as to why your autograding is failing.

**Review by hannahpan1 — CHANGES_REQUESTED** [review]

**Review by hannahpan1 — DISMISSED** [review]
Looks good! Thanks for making all the new changes.

**Pull request merged**
